floors/rule.go

```diff
@@ -0,0 +1,404 @@
+			width = imp.Banner.Format[0].W
+			height = imp.Banner.Format[0].H
+		} else if imp.Banner.W != nil && imp.Banner.H != nil {
+			width = *imp.Banner.W
+			height = *imp.Banner.H
+		}
+	} else {
+		width = imp.Video.W
+		height = imp.Video.H
+	}
+
+	if width != 0 && height != 0 {
+		size = fmt.Sprintf("%dx%d", width, height)
+	}
+	return size
+}
+
+func getDomain(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if request.Site != nil {
+		if len(request.Site.Domain) > 0 {
+			value = request.Site.Domain
+		} else {
+			value = request.Site.Publisher.Domain
+		}
+	} else {
+		if len(request.App.Domain) > 0 {
+			value = request.App.Domain
+		} else {
+			value = request.App.Publisher.Domain
+		}
+	}
+	return value
+}
+
+func getSiteDomain(request *openrtb2.BidRequest) string {
+	var value string
+	if request.Site != nil {
+		value = request.Site.Domain
+	} else {
+		value = request.App.Domain
+	}
+	return value
+}
+
+func getPublisherDomain(request *openrtb2.BidRequest) string {
+	var value string
+	if request.Site != nil {
+		value = request.Site.Publisher.Domain
+	} else {
+		value = request.App.Publisher.Domain
+	}
+	return value
+}
+
+func getBundle(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if request.App != nil {
+		value = request.App.Bundle
+	}
+	return value
+}
+
+func getgptslot(imp openrtb2.Imp) string {
+	var value string
+	adsname, err := jsonparser.GetString(imp.Ext, "data", "adserver", "name")
+	if err == nil && adsname == "gam" {
+		gptSlot, _ := jsonparser.GetString(imp.Ext, "data", "adserver", "adslot")
+		if gptSlot != "" {
+			value = gptSlot
+		}
+	} else {
+		value = getpbadslot(imp)
+	}
+	return value
+}
+
+func extractChanelNameFromBidRequestExt(bidRequest *openrtb2.BidRequest) string {
+	requestExt := &openrtb_ext.ExtRequest{}
+	if bidRequest == nil {
+		return CATCH_ALL
+	}
+
+	if len(bidRequest.Ext) > 0 {
+		err := json.Unmarshal(bidRequest.Ext, &requestExt)
+		if err != nil {
+			return CATCH_ALL
+		}
+	}
+
+	if requestExt.Prebid.Channel != nil {
+		return requestExt.Prebid.Channel.Name
+	}
+	return CATCH_ALL
+}
+
+func getpbadslot(imp openrtb2.Imp) string {
+	value := CATCH_ALL
+	pbAdSlot, err := jsonparser.GetString(imp.Ext, "data", "pbadslot")
+	if err == nil {
+		value = pbAdSlot
+	}
+	return value
+}
+
+func getAdUnitCode(imp openrtb2.Imp) string {
+	adUnitCode := CATCH_ALL
+	gpId, err := jsonparser.GetString(imp.Ext, "gpid")
+	if err == nil && gpId != "" {
+		return gpId
+	}
+
+	if imp.TagID != "" {
+		return imp.TagID
+	}
+
+	pbAdSlot, err := jsonparser.GetString(imp.Ext, "data", "pbadslot")
+	if err == nil && pbAdSlot != "" {
+		return pbAdSlot
+	}
+
+	storedrequestID, err := jsonparser.GetString(imp.Ext, "prebid", "storedrequest", "id")
+	if err == nil && storedrequestID != "" {
+		return storedrequestID
+	}
+	return adUnitCode
+}
+
+func isMobileDevice(userAgent string) bool {
+	isMobile, err := regexp.MatchString("(?i)Phone|iPhone|Android|Mobile", userAgent)
```

Comment:
This regexp doesn't appear to match the requirements, particularly the android patterns:
`'phone' if UA matches one of these patterns: "Phone", "iPhone", "Android.*Mobile", "Mobile.*Android"`